source/bayes_simulation.Rmd

```diff
@@ -45,112 +45,171 @@ correctly identifies 80 percent as "OK" but says that 20 percent are
+   For each car, generate a number between 0 and 1.
 
-2.  Draw a ball. If it is one of those said to be "faulty"---that is,
-    violet or orange---draw (with replacement) another ball. Continue
-    until obtaining a ball said to be "OK"---that is, a blue or maroon
-    ball. Then record its color.
+   - For faulty cars, if the number is between 0 and 0.1, the mechanic says the car is OK.
+   - For working cars, if the number is between 0 and 0.8, the mechanic says the car is OK.
 
-3.  Repeat step 2 perhaps 1000 times and compute the proportions of
-    blues among the recorded results.
+4. Now, we count.  We would like to know how often a car is faulty,
+   out of all the times the car is determined to be OK by a mechanic.  So, we count the number of times that a car is been called OK but is faulty, and divide it by the number of times cars are labeled OK.
 
--OR-
+Here is the code:
 
-1.  Choose a number randomly between "1" and "100."
+```{python}
+import numpy as np
 
-2.  If "28-30" or "31-45:, record; otherwise draw another number, and
-    repeat until a number is recorded.
+N = 1000000
 
-3.  Repeat step 2 and count the proportion "28-30" among the total
-    "28-45."
+dice = np.random.random(N)
+faulty = dice < 0.3
+non_faulty = ~faulty
 
-The key modeling step is excluding a trial from consideration (without
-making any record of it) if it produces an irrelevant observation, and
-continuing to do so until the process produces an observation of the
-sort about which you are presently inquiring.
+dice = np.random.random(N)
+mechanic_says_OK = np.zeros(N, dtype=bool)
+mechanic_says_OK[faulty & (dice < 0.1)] = True
+mechanic_says_OK[~faulty & (dice < 0.8)] = True
 
-Using RESAMPLING STATS, an answer may be produced as follows:
+k = np.count_nonzero(mechanic_says_OK & faulty) / np.count_nonzero(mechanic_says_OK)
 
-"01 - 27" = actually faulty, said to be "faulty" "28 - 30 = faulty, said
-to be "OK"
+print(f'{k * 100:.2}%')
+```
 
-"31 - 86" = OK, "OK"
+The result is around 5.1%.  If we look at the numbers on our diagram, this makes sense: if we started with 100 cars, 59 are declared OK by a mechanic.  Of those, 3 were faulty, and `3/59 = 0.0508`.
 
-"87 - 100" = OK, "faulty"
+### Probability interpretation
 
-`           REPEAT 1000         `
+**Probability from proportion**
 
-Do 1000 repetitions
+In these examples, it is very common for us to calculate proportions.  In a given simulation: How many cars are declared OK by a mechanic?  59/100.  How many of those 59 were faulty?  3/59.  Given evidence of events, we can count how common they were.
 
-`           GENERATE 1 1,100 a         `
+A natural extension of this idea is to predict the probability of an event occurring.  We can say "in past 100 experiments, condition X occurred 20 times, therefore I  estimate that the probability of X occurring, $P(X)$, is 20/100".  The more trial outcomes we examine, the better our probability estimates become.
```

Comment:
Here we get to the idea of a probability being a proportion *over the long run*.  And that a proportion over a short run (like 100) is an estimate of the probability over the long run.  And that the longer the run, the better the estimate.  But this is something serious already, called the [Law of Large Numbers](https://en.wikipedia.org/wiki/Law_of_large_numbers) - and needs some discussion and examples.  Probably this should come in a previous chapter on more basic probability.